File: leetcode/p00145/p00145.py
from typing import Optional, List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        if not root:
            return res
        
        stack = []
        while root or len(stack) > 0:
            while root:
                res.insert(0, root.val)
                stack.append(root)
                root = root.right
            root = stack.pop()
            root = root.left
        
        return res

File: leetcode/p00145/p00145.go
package p00145

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Time Complexity: O(n)
// Space Complexity: O(n)
func postorderTraversal(root *TreeNode) []int {
	res := []int{}
	if root == nil {
		return res
	}

	s := CreateStack()
	for root != nil || !s.Empty() {
		for root != nil {
			res = append(res, root.Val)
			s.Put(root)
			root = root.Right
		}
		root = s.Pop()
		root = root.Left
	}

	head, tail := 0, len(res)-1
	for head < tail {
		res[head], res[tail] = res[tail], res[head]
		head++
		tail--
	}

	return res
}

type Stack struct {
	arr  []*TreeNode
	size int
}

func CreateStack() Stack {
	return Stack{
		arr:  make([]*TreeNode, 0),
		size: 0,
	}
}

func (s *Stack) Empty() bool {
	return s.size == 0
}

func (s *Stack) Put(node *TreeNode) {
	s.arr = append(s.arr, node)
	s.size++
}

func (s *Stack) Pop() *TreeNode {
	if s.size == 0 {
		return nil
	}

	node := s.arr[s.size-1]
	s.arr = s.arr[:s.size-1]
	s.size--

	return node
}
